compiler/luci/pass/src/FuseSliceWithTConvPass.cpp

```diff
@@ -0,0 +1,286 @@
+         indices[1] * node->dim(2).value() * node->dim(3).value() +
+         indices[2] * node->dim(3).value() + indices[3];
+}
+
+luci::Padding compute_padding(const luci::CircleTransposeConv *tconv, int32_t out_height,
+                              int32_t out_width, int32_t pad_top, int32_t pad_left)
+{
+  auto const filter = dynamic_cast<luci::CircleConst *>(tconv->filter());
+  if (!filter)
+    return luci::Padding::UNDEFINED;
+
+  auto tconv_shape = dynamic_cast<luci::CircleConst *>(tconv->inputSizes());
+  if (!tconv_shape)
+    return luci::Padding::UNDEFINED;
+
+  luci::Padding padding = luci::Padding::UNDEFINED;
+  std::initializer_list<luci::Padding> paddings_to_check = {luci::Padding::VALID,
+                                                            luci::Padding::SAME};
+
+  auto const filter_height = filter->dim(1).value();
+  auto const filter_width = filter->dim(2).value();
+  auto const stride_height = tconv->stride()->h();
+  auto const stride_width = tconv->stride()->w();
+
+  for (auto padding_to_check : paddings_to_check)
+  {
+    auto const in_height =
+      compute_input_size(padding_to_check, out_height, filter_height, stride_height);
+    auto const pad_top_virtual =
+      extract_pad_value(stride_height, in_height, filter_height, out_height);
+    if (pad_top_virtual != pad_top)
+      continue;
+
+    auto const in_width =
+      compute_input_size(padding_to_check, out_width, filter_width, stride_width);
+    auto const pad_left_virtual =
+      extract_pad_value(stride_width, in_width, filter_width, out_width);
+    if (pad_left_virtual == pad_left)
+    {
+      padding = padding_to_check; // adissible paddding is found
+      break;
+    }
+  }
+
+  return padding;
+}
+
+/**
+ *  Fuse Slice with CircleTransposeConv if possible
+ *
+ *  NOTE: In case predecessor of slice is tconv, we can try to merge slice with tconv,
+ *  because spatial slice is reduction so as padding for tconv,
+ *  while channels slice reduction can be directly modeled in tconv.
+ *  For now there is no option to set explicitely pad values for
+ *  CircleTransposeConv. Only using VALID/SAME and output shape is the only way
+ *  to set pad values. That is why not all numerical values of pad are legal for such
+ *  transform.
+ *
+ *  BEFORE
+ *                    |
+ *           [CircleTransposeConv]
+ *                    |
+ *               [CircleSlice]
+ *                    |
+ *
+ *  AFTER (if possible)
+ *                    |
+ *            [CircleTransposeConv] (with m.b. changed padding, output shape, and filter/bias)
+ *                    |
+ *
+ */
+
+bool fuse_slice_with_tconv(luci::CircleSlice *slice)
+{
+  // NOTE: assume NHWC layout
+  auto tconv = dynamic_cast<luci::CircleTransposeConv *>(slice->input());
+  RETURN_FALSE_UNLESS(tconv != nullptr);
+
+  auto const tconv_users = loco::succs(tconv);
+  // in general no fusion is possible if tconv is used by many successors
+  RETURN_FALSE_UNLESS(tconv_users.size() == 1);
+
```

Comment:
@jinevening 
As far i understand:
in case all users of `tconv` are different slices, then this pass will produce as many `TConv`s as users. Is it intended?

Reply:
@jinevening
Should i remove this check or change comment?

Reply:
Yes. That's how our passes work in general :) We can limit the case to work just for a single user (e.g., for peformance). Then we may update the comment (the current comment is somewhat misleading).

Reply:
ok

Reply:
done